### ISSUE TITLE
Custom inner block is not rendered on checkout when added manually from block editor

### ISSUE THREAD

**Issue by rashmimalpande** [woothemes/woocommerce#46264]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

The custom inner block is only rendered on the frontend if it is a forced block. It is not possible to let merchants add the custom blocks anywhere on the checkout page.

### Expected behavior

If we remove the 'lock' attribute from block.json file, then it should be possible to add the custom inner block in any of the mentioned parent blocks in block.json file.  

### Actual behavior

The custom inner block is rendered on the frontend only if we make it a forced block by adding the 'lock' attribute. 

### Steps to reproduce

1. Create a custom inner block with an input field. 
2. In the block.json file, the category and parent is set as - 
```
"category": "woocommerce",
"parent": ["woocommerce/checkout-contact-information-block" ]
```
The 'attributes' property does not contain 'lock' attribute.
3. In the block editor, an input field can be added manually in the contact section with the help of registerBlockType
4. On the frontend, the block is registered as -
```
const options = { 
	metadata,
	component: () => {
                  <ValidatedTextInput
				id="input_field"
				type="text"
				required={ false }
				label={
					'Field label'
				}
				value={ '' }
			/>
         }
};

registerCheckoutBlock( options );
```
5. On creating the build and enqueing scripts through the IntegrationInterface, the input field is not displayed on checkout. 
6. If the "lock" attribute is added then the input field is rendered on checkout.

### WordPress Environment ###

WordPress address (URL): http://wccustom.local
Site address (URL): http://wccustom.local
WC Version: 8.7.0
REST API Version: ✔ 8.7.0
Action Scheduler Version: ✔ 3.7.2
Log Directory Writable: ✔
WP Version: 6.5
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 8.2.10
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.85.0
OpenSSL/3.0.8

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.4.0
WC Database Prefix: wp_
Total Database Size: 11.17MB
Database Data Size: 7.13MB
Database Index Size: 4.04MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.08MB + Index: 0.09MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 1.44MB + Index: 0.53MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.31MB + Index: 0.17MB + Engine InnoDB
wp_aioseo_cache: Data: 0.11MB + Index: 0.03MB + Engine InnoDB
wp_aioseo_notifications: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_aioseo_posts: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.06MB + Index: 0.09MB + Engine InnoDB
wp_edd_adjustmentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_adjustments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_edd_customer_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_edd_customer_email_addresses: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_edd_customermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_customers: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_edd_logmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_logs: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_edd_logs_api_requestmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_logs_api_requests: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_logs_file_downloadmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_logs_file_downloads: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_edd_notemeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_notes: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_edd_notifications: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_order_addresses: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_edd_order_adjustmentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_order_adjustments: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_edd_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_order_items: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_edd_order_transactions: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_edd_ordermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_edd_orders: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.03MB + Index: 0.09MB + Engine InnoDB
wp_postmeta: Data: 0.30MB + Index: 0.28MB + Engine InnoDB
wp_posts: Data: 0.16MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.06MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.08MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.11MB + Index: 0.19MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wpmailsmtp_debug_events: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpmailsmtp_tasks_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 23
page: 13
post: 2
product: 18
product_variation: 7
revision: 13
shop_order: 59
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (5) ###

Query Monitor: by John Blackbourn – 3.15.0
WooCommerce: by Automattic – 8.7.0
WP Mail SMTP: by WP Mail SMTP – 3.11.1

### Inactive Plugins (12) ###

All in One SEO: by All in One SEO Team – 4.5.6
Easy Digital Downloads: by Easy Digital Downloads – 3.2.7.1
Easy Digital Downloads - Auto Register: by Easy Digital Downloads – 1.4.4
WooCommerce Checkout Manager: by QuadLayers – 7.1.5

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: ✔

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 5 KB

### WC Pages ###

Shop base: #8 - /shop/
Cart: #9 - /cart/
Checkout: #10 - /checkout/
My account: #11 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.3.0 (update to version 4.5.4 is available)
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-04-06 06:02:20 +05:30
Options: ✔
Notes: 117
Onboarding: completed

### Action Scheduler ###

Complete: 288
Oldest: 2024-03-05 11:49:55 +0000
Newest: 2024-04-05 10:44:07 +0000

Failed: 798
Oldest: 2023-02-22 13:26:06 +0000
Newest: 2024-03-20 05:22:05 +0000

Pending: 4
Oldest: 2024-04-05 11:44:07 +0000
Newest: 2024-04-09 05:15:42 +0000

### Status report information ###

Generated at: 2024-04-05 16:14:16 +05:30
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by pmcpinto** [comment]
@ralucaStan is this issue in progress? I don't see anyone assigned to it.

**Comment by ralucaStan** [comment]
@wavvves unassigned himself and might have missed moving it back to the backlog. Paulo, could you help clarify if this is the case? Thank you

**Comment by wavvves** [comment]
I'm sorry, that was exactly what happened. I assigned myself and moved the wrong ticket, only realized when I opened it and forgot to move after unassigning.

**Comment by opr** [comment]
I've been working on this and have some additional context to share:

- When saving the page, the blocks are saved as HTML.
- When it comes to front-end rendering, the [`renderInnerBlocks` function](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/atomic/utils/render-parent-block.tsx#L129) only considers elements with a `data-block-name` attribute.
- Elements get mapped from these DOM elements to a React component based on this block name during `renderInnerBlocks`
- If an element doesn't have a `data-block-name` attribute, it is returned as-is. (In the case I was testing with, my custom block just rendered an empty div with text inside)
- Elements _get_ this `data-block-name` attribute during [`add_data_attributes`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/BlockTypesController.php#L141).
- Whether an element gets a `data-block-name` attribute depends on its namespace (all `woocommerce` and `woocommerce-checkout` namespaced blocks get one as well as any other namespaces added using the `__experimental_woocommerce_blocks_add_data_attributes_to_namespace` filter). Individual blocks can also be added using the `__experimental_woocommerce_blocks_add_data_attributes_to_block` filter.

So, the next steps for this issue are to determine:

- Why do we impose a limitation on which blocks get rendered?
- Would it be problematic if we allow any block with a `woocommerce/*` block as its parent to be rendered?

My thoughts on the second point are: no it wouldn't be, because block developers can achieve this anyway by using the `__experimental_woocommerce_blocks_add_data_attributes_to_block` filter, it just means they have to use one less step.

It does require the block to be registered in PHP _and_ JS though.

Here is a POC PR that shows how checking the parents might look:

https://github.com/woocommerce/woocommerce/pull/48543

**Comment by opr** [comment]
@rashmimalpande thanks for reporting this, the fix in #48543 should ship with WooCommerce 9.2!
